pkg/addonmanager/controllers/managementaddonconfig/controller.go

```diff
@@ -0,0 +1,339 @@
+			name := configReference.DesiredConfig.Name
+
+			cma.Status.InstallProgressions[i].ConfigReferences[j].DesiredConfig.SpecHash, err = c.getConfigSpecHash(group, resource, namespace, name)
+			if err != nil {
+				return nil
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *clusterManagementAddonConfigController) patchConfigReferences(ctx context.Context, old, new *addonapiv1alpha1.ClusterManagementAddOn) error {
+	if equality.Semantic.DeepEqual(new.Status.DefaultConfigReferences, old.Status.DefaultConfigReferences) &&
+		equality.Semantic.DeepEqual(new.Status.InstallProgressions, old.Status.InstallProgressions) {
+		return nil
+	}
+
+	oldData, err := json.Marshal(&addonapiv1alpha1.ClusterManagementAddOn{
+		Status: addonapiv1alpha1.ClusterManagementAddOnStatus{
+			DefaultConfigReferences: old.Status.DefaultConfigReferences,
+			InstallProgressions:     old.Status.InstallProgressions,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	newData, err := json.Marshal(&addonapiv1alpha1.ClusterManagementAddOn{
+		ObjectMeta: metav1.ObjectMeta{
+			UID:             new.UID,
+			ResourceVersion: new.ResourceVersion,
+		},
+		Status: addonapiv1alpha1.ClusterManagementAddOnStatus{
+			DefaultConfigReferences: new.Status.DefaultConfigReferences,
+			InstallProgressions:     new.Status.InstallProgressions,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	patchBytes, err := jsonpatch.CreateMergePatch(oldData, newData)
+	if err != nil {
+		return fmt.Errorf("failed to create patch for addon %s: %w", new.Name, err)
+	}
+
+	klog.V(4).Infof("Patching addon %s/%s config reference with %s", new.Namespace, new.Name, string(patchBytes))
+	_, err = c.addonClient.AddonV1alpha1().ClusterManagementAddOns().Patch(
+		ctx,
+		new.Name,
+		types.MergePatchType,
+		patchBytes,
+		metav1.PatchOptions{},
+		"status",
+	)
+	return err
+}
+
+func (c *clusterManagementAddonConfigController) getConfigSpecHash(group, resource, namespace, name string) (string, error) {
+	lister, ok := c.configListers[toListerKey(group, resource)]
+	if !ok {
+		return "", nil
+	}
+
+	var config *unstructured.Unstructured
+	var err error
+	if namespace == "" {
+		config, err = lister.Get(name)
+	} else {
+		config, err = lister.Namespace(namespace).Get(name)
+	}
+
+	if errors.IsNotFound(err) {
+		return "", nil
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	return GetSpecHash(config)
+}
+
+func getDefaultConfigIndex(config addonapiv1alpha1.DefaultConfigReference) string {
+	if config.DesiredConfig.Namespace != "" {
+		return fmt.Sprintf("%s/%s/%s/%s", config.Group, config.Resource, config.DesiredConfig.Namespace, config.DesiredConfig.Name)
+	}
+
+	return fmt.Sprintf("%s/%s/%s", config.Group, config.Resource, config.DesiredConfig.Name)
+}
+
+func getInstallConfigIndex(config addonapiv1alpha1.InstallConfigReference) string {
+	if config.DesiredConfig.Namespace != "" {
+		return fmt.Sprintf("%s/%s/%s/%s", config.Group, config.Resource, config.DesiredConfig.Namespace, config.DesiredConfig.Name)
```

Comment:
i think the index for empty namespace could "foo/bar//name". This is still unique for each one, right?

Reply:
yes, still unique for each one, to check if namespace is empty or not is to make it easy to use `cache.DeletionHandlingMetaNamespaceKeyFunc(obj)` to get above `namespace/name` or `name` I think.

Reply:
to may understanding DeletionHandlingMetaNamespaceKeyFunc can also handle /name as name, the returned namespace is just empty